### ISSUE TITLE
🐸  TTS roadmap

### ISSUE THREAD

**Issue by erogol** [coqui-ai/TTS#378]
These are the main dev plans for :frog: TTS. 

If you want to contribute to :frog: TTS and don't know where to start you can pick one here and start with our [Contribution Guideline](https://github.com/coqui-ai/TTS#contribution-guidelines). We're also always here to help. 

Feel free to pick one or suggest a new one. 

Contributions are always welcome :muscle:  .

## v0.1.0 Milestones
- [x] Better model config handling #21 
- [x] TTS recipes for public datasets.
- [x] TTS trainer API to unify all the model training scripts.
- [x] TTS, Vocoder and SpeakerEncoder model abstractions and APIs. 
- [x] Documentation for 
   - [x] Implementing a new model using :frog: TTS.
   - [x] Training a model on a new dataset from gecko. 
   - [x] Using ```Synthesizer``` interface on ```CLI``` or ```Server```.
   - [x] Extracting Spectrograms for Vocoder training. 
   - [x] Contributing a new pre-trained :frog: TTS model.
   - [x] Explanation for Model config parameters/

## v0.2.0 Milestones
- [x] Grapheme 2 Phoneme in-house conversion. (Thx to gruut  👍 )
- [x] Implement VITS model.

## v0.3.0 Milestones
- [x] Implement generic ForwardTTS API.
- [x] Implement Fast Speech model. 
- [x] Implement Fast Pitch model. 

## v0.4.0 Milestones
- [x] Trainer API v2 - [join the discussion](https://github.com/coqui-ai/TTS/discussions/840)
- [x] Multi-speaker VCTK recipes for all the `TTS.tts` models.

## v0.5.0 Milestones
- [x] Support for multi-lingual models 
- [x] YourTTS release 🚀

## v0.6.0 Milestones
- [x] Add ESpeak support
- [x] New Tokenizer and Phonemizer APIs #937 
- [x] New Model API #1078 
- [x] Splitting the trainer as a separate repo [👟Trainer](https://github.com/coqui-ai/Trainer)
- [x] Update VITS model API
- [x] Gradient accumulation. #560 (in 👟)

## v0.7.0 Milestones
- [x] Implement Capacitron 👑 @a-froghyar 👑 @WeberJulian 
- [x] Release pretrained Capacitron

## v0.8.0 Milestones
- [x] Separate numpy transforms 
- [x] Better data sampling for VITS 
- [x] New Thorsten DE models 👑 @thorstenMueller

## 🏃‍♀️ Milestones along the way
- [ ] Implement End-to-end training API for ForwardTTS models a vocoder. #1510 
- [ ] Implement a Python voice synthesis API.
- [ ] Inject phonemes to the input text at inference. #1452 
- [ ] AdaSpeech1/2 https://arxiv.org/pdf/2104.09715 and https://arxiv.org/abs/2103.00993
- [ ] Let the user pass a custom text cleaner function.  
- [ ] Refactor the text cleaners for a more flexible and transparent API. 
- [ ] Implement [HifiGAN2](https://pixl.cs.princeton.edu/pubs/Su_2021_HSS/Su-HiFi-GAN-2-WASPAA-2021.pdf) (not the vocoder) 
- [ ] Implement emotion and style adaptation. 
- [ ] Implement FastSpeech2 (https://arxiv.org/abs/2006.04558). 
- [ ] [AutoTTS 🤖](https://github.com/coqui-ai/TTS/pull/696) (👑 @loganhart420)
- [ ] Watermarking TTS outputs to sign against DeepFakes.
- [ ] Implement SSML v0.0.1 
- [ ] ONNX and TorchScript model exports. 
- [ ] TensorFlow run-time for training models.

## 🤖 New TTS models 
- [x] AlignTTS (@erogol) 
- [x] HiFiGAN (#16 :crown: @rishikksh20 and @erogol)
- [x] UnivNet Vocoder ( :crown: @rishikksh20)
- [x] VITS [paper](https://arxiv.org/abs/2106.06103)
- [x] FastPitch [source](https://fastpitch.github.io/)
- [x] Alignment Network [paper](https://arxiv.org/abs/2108.10447) 
- [x] End2End TTS combining aligner + tts + vocoder. 
- [x] Multi-Lingual TTS (#11 :crown: @WeberJulian )
- [ ] ParallelTacotron [paper](https://arxiv.org/pdf/2103.14574.pdf) (open for contribution)
- [ ] Efficient TTS [paper](https://arxiv.org/pdf/2012.03500.pdf) (open for contribution)
- [ ] Gaussian length regulator from https://arxiv.org/pdf/2010.04301.pdf (open for contribution)
- [ ] LightSpeech from https://arxiv.org/pdf/2102.04040.pdf (open for contribution)
- [ ] AdaSpeech1/2 https://arxiv.org/pdf/2104.09715 and https://arxiv.org/abs/2103.00993

**Comment by lucascassiano** [comment]
great project! Excited to see this growing!

**Comment by AndrewBarfield** [comment]
I'm learning the code/API and performing experiments. I hope to contribute soon.

I'm also wondering if I can donate (money) to Coqui?

**Comment by kdavis-coqui** [comment]
> I'm learning the code/API and performing experiments. I hope to contribute soon.
> 
> I'm also wondering if I can donate (money) to Coqui?

Wow! Thanks! Humbling.

We were setting up GitHub sponsors, but the tax implications were onerous.

We're currently exploring Patreon. So stay tuned!

**Comment by agrinh** [comment]
@erogol Thanks for sharing the plans!

Do you have any thoughts (or need help to) simplifying the dependencies a bit? I'm thinking that if `TTS` is used as a lib installed over `pip` it might be nice to remove visualisation dependencies only used in notebooks, removing test/dev dependencies and moving e.g. `tensorflow` into extras to reduce the footprint. Personally would love to use this as a dependency rather than maintaining my own fork.

**Comment by erogol** [comment]
@agrinh Why do you need to keep your own fork exactly? It'd be better to expand the conversation on gitter if you like.

**Comment by agrinh** [comment]
> @agrinh Why do you need to keep your own fork exactly? It'd be better to expand the conversation on gitter if you like.

Wow, thanks for the super fast reply. Sure, we can move the discussion to gitter.

**Comment by Sadam1195** [comment]
Please add DC-TTS to the the list of models. 

**DC-TTS** implementation available with MIT Licence [code available here](https://github.com/tugstugi/pytorch-dc-tts)
EFFICIENTLY TRAINABLE TEXT-TO-SPEECH SYSTEM BASED ON DEEP CONVOLUTIONAL NETWORKS WITH GUIDED ATTENTION [paper](https://arxiv.org/pdf/1710.08969.pdf)
@erogol

**Comment by will-rice** [comment]
What were you thinking about the "TensorFlow run-time for training models"? Like giving the user the option of using TensorFlow or PyTorch? I wouldn't mind taking a stab at the TensorFlow part.

**Comment by erogol** [comment]
@will-rice  the plan is to mirror what we have in torch to TF as much as possible. It'd be great if you initiate the work

**Comment by lucashueda** [comment]
Are you guys planning to develop some expressive TTS architectures? I'm currently studying this topic and planning to implement some of them based on Coqui, part of them just controlling latent space using GST [Kwon et al 2020](https://www.researchgate.net/publication/334756998_An_Effective_Style_Token_Weight_Control_Technique_for_End-to-End_Emotional_Speech_Synthesis)  or RE [Sorin et al 2020](https://www.isca-speech.org/archive/interspeech_2020/sorin20_interspeech.html), and others that actually changes the architecture by adding VAE, normalizing flows and gradient reversal

**Comment by a-froghyar** [comment]
@lucashueda Capacitron VAE: https://github.com/coqui-ai/TTS/pull/510

**Comment by lucashueda** [comment]
> @lucashueda Capacitron VAE: #510

Oh nice, hope to see Capacitron integrated soon. So maybe, in the future I'll be able to contribute with some others expressive architectures

**Comment by BillyBobQuebec** [comment]
@erogol  Look forward to new End-to-End models being implemented, specfically Efficient-TTS! if the paper is accurate, it should blow most 2 stage configurations out of the water, considering it seems to have higher MOS than tacotron2+hifigan, while also seeming to have significantly faster speed than glowtts+fastest vocoder!

**Comment by erogol** [comment]
@BillyBobQuebec I don't think I will implement these models anytime soon. But as they stand, contributions are welcome

**Comment by WeberJulian** [comment]
@BillyBobQuebec but you can try VITS which is close to what you're describing :)

**Comment by BillyBobQuebec** [comment]
> @BillyBobQuebec but you can try VITS which is close to what you're describing :)

Agreed, I am currently trying VITS actually, I have some issues training with the coqui implementation unfortunately, I've posted the issue about the bug today and hope I can get it resolved.

**Comment by hemath1001** [comment]
Hi there! Thanks for your great work! I'm looking forward to training YourTTS on other languages. Will training and fine-tuning code of YourTTS be published soon? I would be very grateful if you could tell me an approximate time~ Have a nice day :-D

**Comment by kerryeon** [comment]
Hello, thanks for great works! I'm a fan of `Coqui TTS`.

I'm porting some of the stuffs in the project to the `Rust` for the following reasons.

* Predictable Performance
* Static-typed Metadata & Model Management
* Multithreaded Server Implementation
* *Just I love Rust*

The `VC` in the `YourTTS` has been successfully implemented. And for this purpose, an example of saving/loading a pretrained `Vits` model has been added in the repo. I write it on Milestones PR because I think my work can be helpful to others :)

* Repository (*RusTTS*): https://github.com/kerryeon/rustts

**Comment by erogol** [comment]
@kerryeon great work!! Thanks for sharing!

**Comment by paolo-caroni** [comment]
Any plan to a port of coqui-ai engine for android? TTS on android is very robotic (espeak, rhvoice, festival lite).

**Comment by erogol** [comment]
No immediate plans on that

**Comment by Darth-Carrotpie** [comment]
Thumbs up for planning ONNX support. Hope it gets prioritized more!

**Comment by erogol** [comment]
@Darth-Carrotpie what is your use-case of ONNX? (Just want to get some feedback)

**Comment by lexkoro** [comment]
> @Darth-Carrotpie what is your use-case of ONNX? (Just want to get some feedback)

Personally, for me it sounds like a good way to develop Windows nativ TTS applications without needing a Python runtime and/or the big dependencies like pytorch. 

I tried exporting the VITS model to onnx before, but I didn't succeed.
But there are also other obstacles beside executing the model, like phonemization. ^^

Currently I am using [pythonnet](https://github.com/pythonnet/pythonnet) to embed the required python functions directly in my C# code. For Python I use the embedded version to make the App distributable.

**Comment by Darth-Carrotpie** [comment]
@erogol I am trying to run models in Unity. It's environment is in C#, .NET Standard 2.1. Having a universal format model, means in the long run I can not only run models in Windows, but also build the project for other platforms, like Linux. Of course things like tokenization and phonemization are additional hurdles, but if there are open source examples it's quite doable. For models needing tokenizers I've been using BlingFire succesfully, so I reckon there's similar phonemizer helpers / libraries for other languages beside python, including C#.

**Comment by erogol** [comment]
@Darth-Carrotpie run in unity means in the code or integrate it to Unity editor?

Also better to move this to a separate post under the Discussions

**Comment by Darth-Carrotpie** [comment]
> @Darth-Carrotpie run in unity means in the code or integrate it to Unity editor?
> 
> Also better to move this to a separate post under the Discussions

Created a topic on ONNX at Discussions: https://github.com/coqui-ai/TTS/discussions/1479

**Comment by erogol** [comment]
@desh-woes there is no flutter package, unfortunately.

Can you DM me on Gitter or Element (out chat rooms) if you're willing to work on a particular thing?

**Comment by omkarade** [comment]
how train model using word embedding as input

**Comment by erogol** [comment]
@omkarade no support for that yet.

**Comment by WeberJulian** [comment]
> I want to train a custom Your TTS model on my data set. Can you please share me detailed process.

You can read the relevant documentation here: https://tts.readthedocs.io/en/latest/finetuning.html
Also this is the roadmap thread, please ask for support [here ](https://gitter.im/coqui-ai/TTS?utm_source=share-link&utm_medium=link&utm_campaign=share-link) or open a new discussion/issue

**Comment by kerryeon** [comment]
> @desh-woes there is no flutter package, unfortunately.
> 
> Can you DM me on Gitter or Element (out chat rooms) if you're willing to work on a particular thing?

`rustts` author here. [In my repository,](https://github.com/ulagbulag-village/rustts/blob/master/assets/vits/converter.ipynb) there is a example porting from `PyTorch` to `ONNX`.

Also, I have been developing and managing Flutter projects for several years. If you need help, I am willing to help!

However, it seems that some consensus with the current project is necessary, and I think an agreement is needed.

**Comment by windowshopr** [comment]
Looking forward to the SSML implementation!

**Comment by Kthulu120** [comment]
@erogol is the NaturalSpeech paper something you'd think about implementing I could take a crack at it.

**Comment by erogol** [comment]
@Kthulu120 sure thing. Feel free to shoot a PR. We are always here to help.

**Comment by EricTulowetzke** [comment]
Will there be a C API to this library like your STT library?

**Comment by erogol** [comment]
Not in the roadmap currently

**Comment by JediMaster25** [comment]
I've just tried this with a [random article](https://nautil.us/the-trouble-with-the-big-bang-238547/). Firefox reader view, select all, copy, paste to file,
```
cat "text" | sentences | xargs -0 tts --model_name "tts_models/en/ljspeech/tacotron2-DDC"  --out_path "out.wav" --text
```
and I got a failure `max_decoder_steps`. After adding it `max_decoder_steps: 20000` I got another failure
```
RuntimeError: Calculated padded input size per channel: (4). Kernel size: (5). Kernel size can't be greater than actual input size
```
It takes 6 minutes processing some text to give an error and doesn't even return the sound it has processed up to that point.

I hope there is a better alternative in a couple years because this doesn't seem usable.

**Comment by JediMaster25** [comment]
I hadn't seen that [someone made an script](https://github.com/coqui-ai/TTS/discussions/1101#discussioncomment-2304472) to make using this program with text files simpler.

**Comment by erogol** [comment]
@JediMaster25 you think the Roadmap is the right place for this convo?

**Comment by NicoleKai** [comment]
It's good to see progress on a propper tts project. I'm running arch and no cuda and I am gonna see if I can convince it to use my cpu instead! 
What would be really cool would be if this could work on AVX512 in amdchipsets.

**Comment by erogol** [comment]
I close this as it is hard to follow and automate.

**Comment by nfaraji2002** [comment]
Hi
thanks for delightful codes!
I want to use this version of TTS on raspberry pi 4, but I think this version does not support real time processing.
Are there TF utilities provided as in Mozilla TTS to convert trained models to tf-lite?
Can the strategy of quantization work here for real-time processing?
I need some roadmaps in this regard.

Thanks
Neda

**Comment by jhj0517** [comment]
Thank you for your great work for TTS.

Is there any progress on  `Let the user pass a custom text cleaner function.` ? 
If it's possible, I want to pass my own Korean cleaners.

**Comment by erogol** [comment]
You can currently do it by creating your own tokenizer or overloading the class.

**Comment by MaxIakovliev** [comment]
Marvelous project.
Any ways to donate to core contributors?
I would prefer to use paypal.

**Comment by erogol** [comment]
@MaxIakovliev you can use https://coqui.ai/ :)

**Comment by erogol** [comment]
This roadmap issue is quite outdated. I'll keep it open to keep the references to some of the issues and models we like to tackle but won't be updating until one day officially becomes 48 hours.

**Comment by jmlcoliveira** [comment]
Any update regarding SSML implementation?

**Comment by erogol** [comment]
We are not working on SSML currently, it is back in the list without a precise timeline.

**Comment by offside609** [comment]
Please do!!

**Comment by violet17** [comment]
Will you support bark-small? Thanks.

**Comment by csukuangfj** [comment]
> Any plan to a port of coqui-ai engine for android? TTS on android is very robotic (espeak, rhvoice, festival lite).

@paolo-caroni 



Please take a look at
https://github.com/coqui-ai/TTS/discussions/3194

You can use [sherpa-onnx](https://github.com/k2-fsa/sherpa-onnx) to run VITS models from Coqui on Android and also embedded devices, e.g., raspberry pi.


We have pre-built Android APKs for the VITS English models from Coqui.
https://k2-fsa.github.io/sherpa/onnx/tts/apk.html

![image](https://github.com/coqui-ai/TTS/assets/5284924/56b2ec37-27de-40c0-8dd4-8245ae2ac8be)

**Comment by DmitryVN** [comment]
Fix it plz https://github.com/coqui-ai/TTS/issues/3039  https://github.com/coqui-ai/TTS/issues/3282
The problem persists and because of this, normal correct use is not possible. Also at the moment it kind of breaks off the phrase at the end of each sentence and it turns out a jerky reading.

**Comment by MarkChrisE2091** [comment]
Any new update?

**Comment by MarkChrisE2091** [comment]
Any new update?

**Comment by csukuangfj** [comment]
> Any plan to a port of coqui-ai engine for android? TTS on android is very robotic (espeak, rhvoice, festival lite).

@paolo-caroni 

We have supported it in https://github.com/k2-fsa/sherpa-onnx/pull/508

The following is a YouTube video
https://www.youtube.com/watch?v=33QYuVzDORA

You can use all coqui-ai/TTS models and piper models listed in
https://github.com/k2-fsa/sherpa-onnx/releases/tag/tts-models
with https://github.com/k2-fsa/sherpa-onnx/pull/508

**Comment by imevro** [comment]
hi guys, why?

upd: found https://twitter.com/_josh_meyer_/status/1742522906041635166

<img width="588" alt="Screenshot 2024-04-29 at 19 19 25" src="https://github.com/coqui-ai/TTS/assets/775692/b9bedf00-0be1-4d43-936f-2a529a0db50d">

**Comment by NicoleKai** [comment]
Their ability to exist and be profitable was dependent on how much better their tech was compared to everyone else. It may not feel like it, but we are in the middle of an AI singularity. Coqui's business model might have stood a chance if they started with this tech 5 years earlier, but it was probably too little too late. Eleven labs is probably eating their lunch :/

**Comment by angadkwatra73** [comment]
Can we still use their product well, even though they have shut down?

**Comment by eginhard** [comment]
> Can we still use their product well, even though they have shut down?

You can still use the code and models (within the license terms). We also maintain [a fork](https://github.com/idiap/coqui-ai-TTS) that has a bunch of bug fixes and other small updates.
